### ISSUE TITLE
Consistent asset compilation complaint with certain seeds

### ISSUE THREAD

**Issue by pond** [rubycdp/cuprite#186]
We have a large proprietary application recently converted from Rails 6 to 7. The RSpec-based test suite is huge. It was using Capybara/Webdrivers for tests originally, but when we upgraded to Rails 7, we changed to Cuprite/Ferrum.

To explain the issue we're seeing with Cuprite, first, some history:

* We use AWS CI and had noticed under Rails 6 with Webdrivers 5, that _all_ of our download tests would fail. We were unable to replicate this locally. It looked like Chrome was, specifically and only under CI, ignoring the requested download location so that the test suite would never find an awaited file. We have a Dockerised CI setup and this wasn't even replicable when running the exact container on local machine, so it was very mysterious. After burning a lot of hours with no solution, we gave up and stayed with Webdrivers 4.

* The move to Rails 7 was difficult requiring a _lot_ of work to get things going reliably. Turbolinks->Turbo is a huge headache. We got it going, but unfortunately found that the aforementioned download spec failure now _always_ happened, regardless of Webdrivers version. Then I read about Cuprite and thought I'd give it a go - Webdrivers has always felt kinda janky and slow, so, why not! This has been overall positive, even if it initially caused a lot of fragile test flickering, not just on race conditions here and there but just things like Cuprite seeming to have a different idea of what "visible" means or, sometimes, simply seeming to not be able to talk reliably to Chrome, but the download specs all worked, our system suite time-to-pass reduced from around 40 minutes to 30 minutes which is awesome, and we eventually got things settled to a reasonable degree.

So, mostly things are a success - **except** for one persistent serious issue. On some seeds, we get the following error when the first system spec tries to run:

```
          ActionView::Template::Error:
            Asset `application.css` was not declared to be precompiled in production.
            Declare links to your assets in `app/assets/config/manifest.js`.
          
              //= link application.css
          
            and restart your server
```

The pipeline is working fine in all environments usually, including in test; but just for _some seeds_ we get this error. It seems to happen if the very first system test runs using Cuprite, but if any system test prior uses the Rack Driver (`js: false`, basically) then the exception above does _not_ occur at all and tests run normally.

We do not use and have never moved to Webpacker, so just use vanilla Sprockets with very light-touch use of JavaScript via jQuery and Bootstrap. **We do use SASS** and therefore our "entrypoint" for styles is actually under filename `application.scss`.

Google gives me no results! I am a loss to explain it; I _can_ replicate this reliably on localhost with just `bundle exec rake tmp:clear` then executing system tests on the relevant seed, but I don't see how or why it should arise. Adding the suggested link line to the manifest is of course not required given the usual manifest setup and, if added anyway, unsurprisingly resolves nothing.

* Trying `true`/`false` in `config.assets.compile` in `config/environments/test.rb` does not help; I've generated a new, clean Rails 7 application and compared environment and other configuration template files line by line and can't see what might be wrong.

* Astonishingly, disabling the raising of exceptions for missing assets via `config.assets.check_precompiled_asset = false` does, as expected, suppress the exception - but even though whatever is going wrong must still be going wrong "under the hood", all tests then appear to run normally and pass. Dropping a `screenshot_and_open_image` call into the very first of the Cuprite tests that runs shows a normal rendered page with all of the expected styles present and correct.

* `bundle update` says we're on the "latest everything" - e.g. Rails 7.0.2.2, though we've seen this issue for all of Rails 7.*, Cuprite 0.13, Ferrum 0.11, Sprockets 4.0.2, Sprockets-Rails 3.4.2, RSpec-Rails 5.1, RSpec-Core 3.1. We run on Ruby 3.1.0p0 and have not enabled the experimental JIT.

Anyone have any ideas what might be going on here?

**Comment by route** [comment]
Hm that's crazy, did you try compiling assets beforehand? Even before running the test suite?

**Comment by pond** [comment]
@route No, it should never be necessary to precompile assets in test or development; they're only precompiled in production because on-demand compilation is considered too expensive an operation.

Given that the only reliable way to replicate is to clear `tmp` between test runs, I have little doubt that precompilation would work, but it should not be needed and, even stranger, ignoring the compilation error doesn't stop the first test from running normally with all assets in place as expected.

**Comment by route** [comment]
You are saying that if the first test is Cuprite test it fails, does it mean if we clear tmp dir and run test intentionally like `rspec spec/features/...` we'll see error?

**Comment by pond** [comment]
I don't know, but perhaps. I know we see it in _our_ application. Since nobody has engaged with this issue to say "yes, it's <x>", or even to say they'd seen it and worked around it too, I think I'll have to try and make a replicable test case. Obviously that might take a while as I'm starting from a giant proprietary application and can't leave any "private" IP in there `:-)`

**Comment by pond** [comment]
In the end we dropped the Sprockets workflow entirely, gave up and just moved to an all-Node workflow with `jsbundling-rails` and `cssbundling-rails`. It wasn't really tenable to keep using Sprockets and import maps don't provide a CSS solution, so we had to just take the pain. As a result, the above problem goes away as we have no compilation stages left in the pipeline that aren't explicitly activated by extensions to our RSpec startup script data.

See https://github.com/rails/importmap-rails/issues/107#issuecomment-1049713670 for rationale for the switchover and https://stackoverflow.com/a/71289436 for the RSpec solution we wrote.

Thanks for your assistance `:-)` - may as well close this now.
